File: Tree/527-hasPathSum-bfs.cpp
#include<iostream>
#include<queue>

using std::cout;

class TreeNode{
    public:
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode(int v,TreeNode* l=nullptr, TreeNode* r=nullptr):val{v},left{l},right{r}{};
};
using Q=std::queue<std::pair<const TreeNode*,int>>;
bool hasPathSum(const TreeNode* root,int target)
{
    if(!root)
        return false;

    Q q{};
    q.push(std::make_pair(root,0));
    while(!q.empty())
    {
        std::pair<const TreeNode*,int> top=q.front();
        q.pop();
        const TreeNode* curNode=top.first;
        int s=top.second;
        int curVal=s+curNode->val;
        if((curVal==target)&&(!curNode->left)&&(!curNode->right))
            return true;
        if(curNode->left)
            q.push(std::make_pair(curNode->left,curVal));
        if(curNode->right)
            q.push(std::make_pair(curNode->right,curVal));
    }
    return false;
    
}

int main()
{
    TreeNode* root=new TreeNode(7);
    TreeNode* one=new TreeNode(1);
    TreeNode* two=new TreeNode(2);
    TreeNode* three=new TreeNode(2);

    root->left=one;
    one->left=two;
    one->right=three;
    cout<<hasPathSum(root,11);
}

File: Tree/527-hasPathSum-bfs.py

from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
def hasPathSum(root,target):
    if not root:
        return false;
    q=deque([(root,0)])  #here can not be deque(((root,0)))
    while q:
        curNode,s=q.pop()
        curVal=curNode.val+s
        if curVal==target and not curNode.left and not curNode.right:
            return True
        if curNode.left:
            q.append((curNode.left,curVal)) 
        if curNode.right:
            q.append((curNode.right,curVal))
    return False


root=TreeNode(7)
left1=TreeNode(1)
left12=TreeNode(2)
right12=TreeNode(5)
root.left=left1
left1.left=left12
left1.right=right12

print(hasPathSum(root,11))
    